### PR TITLE
[WIP] Move to NetStandard

### LINKED ISSUE #146: Drop support for EnterpriseLibrary.Logging
Version 5.0 which will be netstandard / netcoreapp support only. The official [EntperiseLibrary.Logging](https://www.nuget.org/packages/EnterpriseLibrary.Logging/6.0.1304) package hasn't had a release in over 4 years and the source now lives in the [MicrosoftArchive](https://github.com/MicrosoftArchive/enterprise-library) github organization. 

There is a [community fork](https://github.com/Chavoshi/EnterpriseLibrary.NetCore) however with 1 contributor, 3 stars and 9 watchers I assert this to not be popular enough to support. With the advent of `Microsoft.Extensions.Logging.Abstractions 2.0.0`. 

If anyone disagrees, please let me know.

### LINKED ISSUE #147: Drop support for PCL
Netstandard replaces this and PCL projects can still use version 4.x

### LINKED ISSUE #148: Replace Gibraltar.Agent with Loupe.Agent.Core
@kendallmiller will there be a `netstandard` version of Gibraltar.Agent?

For now I'm going to exclude the code and the tests.

### PR DESCRIPTION
**Pull request by damianh**
- [x] Convert all projects to new csproj format
 - [x] Remove PLC support and all `LIBLOG_PORTABLE` compilation directive. (fixes #147 )
 - [x] Remove support for EnterpriseLibrary.Logging (fixes #146 )
 - [x] Update Serilog, NLog, Log4net & Gibraltar.Agent packages that support `netstandard` (and will only support those versions or later). (fixes #148)
 - [x] Remove special test projects for older multiple Serilog and NLog
 - [x] ~Multi-target netstandard1.0 through 2.0, netcoreapp1.0 through 2.0 and net45~ Just targeting `netstandard2.0`
 - [ ] ~Update cake build to use Cake.CoreCLR~ Replaced it with plain PS script.
 - [x] Fix up nuspec to create content package, and test it!
 - [x] Bump to version 5.0.0

### CODE REVIEW

File: src/LibLog/LibLog.cs [hunk 1]
```diff
@@ -1368,15 +1330,16 @@ private static Type GetLogManagerType()
...
             Type logManagerType = GetLogManagerType();
-            MethodInfo method = logManagerType.GetMethodPortable("GetLogger", typeof(string));
+            var log4netAssembly = Assembly.GetAssembly(logManagerType);
+            MethodInfo method = logManagerType.GetMethodPortable("GetLogger", typeof(Assembly),  typeof(string));
+            ParameterExpression repositoryAssemblyParam = Expression.Parameter(typeof(Assembly), "repositoryAssembly");
             ParameterExpression nameParam = Expression.Parameter(typeof(string), "name");
-            MethodCallExpression methodCall = Expression.Call(null, method, nameParam);
-            return Expression.Lambda<Func<string, object>>(methodCall, nameParam).Compile();
+            MethodCallExpression methodCall = Expression.Call(null, method, repositoryAssemblyParam, nameParam);
+            var lambda =  Expression.Lambda<Func<Assembly, string, object>>(methodCall, repositoryAssemblyParam, nameParam).Compile();
```
Comment (damianh) [inline src/LibLog/LibLog.cs]: I believe this will fix #128

**Comment by snakefoot** [comment]
@damianh Notice that many of the LIBLOG_PORTABLE are required for NETSTANDARD to work (Reflection stuff goes through TypeInfo instead of Type before NetStandard2.0)

Sadly enough there is no single NETSTANDARD-flag or COREAPP-flag

.NET Standard 1.0 --> NETSTANDARD1_0
.NET Standard 1.1 --> NETSTANDARD1_1
.NET Standard 1.2 --> NETSTANDARD1_2
.NET Standard 1.3 --> NETSTANDARD1_3
.NET Standard 1.4 --> NETSTANDARD1_4
.NET Standard 1.5 --> NETSTANDARD1_5
.NET Standard 1.6 --> NETSTANDARD1_6
NETCOREAPP1_0
NETCOREAPP1_1

For NetStandard2.0 and NetCoreApp2.0 then there is no need for special defines

**Comment by snakefoot** [comment]
@damianh Instead of using NLog 5.0, then please use NLog 4.5 BETA for NetStandard support.

**Comment by damianh** [comment]
> Instead of using NLog 5.0 BETA, then please use NLog 4.5 BETA for NetStandard support.

Sure. I checked nuget.org and assumed the 5.x line was the one with the netstandard support.

> Notice that many of the LIBLOG_PORTABLE are required for NETSTANDARD to work
> For NetStandard2.0 and NetCoreApp2.0 then there is no need for special defines

This PR targets `netstandard2.0` _to get things going_. I haven't finalized what the minimal netstandard version to support. I'm hoping it's netstandard2.0. I'll open an issue to ask the community.

**Comment by snakefoot** [comment]
> I haven't finalized what the minimal netstandard version to support

It should not be difficult to support all versions. Just have to re-introduce some of the reflection-hacks from PCL. Then it is just a matter of having this at the top of the LibLog.cs file:

```
#if !NETSTANDARD
#if NETSTANDARD1_0 || NETSTANDARD1_1 || NETSTANDARD1_2 || NETSTANDARD1_3 || NETSTANDARD1_4 || NETSTANDARD1_5 || NETSTANDARD1_6 || NETCOREAPP1_0 || NETCOREAPP1_1
#define NETSTANDARD
#endif
#endif
```

**Comment by snakefoot** [comment]
> Sure. I checked nuget.org and assumed the 5.x line was the one with the netstandard support.

It took long time for NLog 5.0 BETA to reach a working state with NetStandard1.5. Then NetStandard2.0 suddenly arrived and made it "easy" to convert NLog 4.4 (which then became NLog 4.5 BETA).

NLog 4.5 BETA also supports structured logging.

**Comment by damianh** [comment]
> Then it is just a matter of having this at the top of the LibLog.cs file:

Dude, that's great! Will give it a spin.

**Comment by snakefoot** [comment]
If you feel that GetCurrentClassLogger could be nice in NetStandard, then it can be implemented with this logic (Using Environment.Stack -> https://apisof.net/catalog/System.Environment.StackTrace)

```
            int framesToSkip = 1;
            string className = string.Empty;
            var stackTrace = Environment.StackTrace;
            var stackTraceLines = stackTrace.Replace("\r", "").Split(new[] { "\n" }, StringSplitOptions.RemoveEmptyEntries);
            for (int i = 0; i < stackTraceLines.Length; ++i)
            {
                var callingClassAndMethod = stackTraceLines[i].Split(new[] { " ", "<>", "(", ")" }, StringSplitOptions.RemoveEmptyEntries)[1];
                int methodStartIndex = callingClassAndMethod.LastIndexOf(".", StringComparison.Ordinal);
                if (methodStartIndex > 0)
                {
                    // Trim method name. 
                    var callingClass = callingClassAndMethod.Substring(0, methodStartIndex);
                    // Needed because of extra dot, for example if method was .ctor()
                    className = callingClass.TrimEnd('.');
                    if (framesToSkip != 0 && !className.StartsWith("System.Environment"))
                    {
                        i += framesToSkip - 1;
                        framesToSkip = 0;
                        continue;
                    }
                    if (!className.StartsWith("System."))
                        break;
                }
            }
            return className;
```

**Comment by damianh** [comment]
@snakefoot `#if !LIBLOG_NETSTANDARD...` worked fine. Thanks for that. Code compiles across all the TFMs.

**Comment by damianh** [comment]
The GetCurrentClassLogger looks interesting and hardcore :) I'd probably add some sort of cache to that. What to send a PR to this branch to get you name stamped on the commit / contrib? If not, I'll pick it up later this week. My weekend hacking is coming to a close...

**Comment by damianh** [comment]
Tests passing (except in ncrunch heh):

**Comment by snakefoot** [comment]
> The GetCurrentClassLogger looks interesting and hardcore :). I'd probably add some sort of cache to that. 

Hard to perform caching of the current StackTrace, and one seldom create new loggers. Usually they are static objects, unless doing dependency-injection where current-class-type has to be provided manually.

**Comment by snakefoot** [comment]
> Tests passing

Please make a search for `NLOG4` in the source code, and remove all `#if `, so the code is always activated. Curious if callsite works for all NetCoreApps

**Comment by damianh** [comment]
Agreed. The DI use case just needs a little thought.

The great thing though, is that after moving TFMs and updating all the logging frameworks, the only break was log4net's `ILoggerRepository` dependency.

**Comment by snakefoot** [comment]
> The great thing though, is that after moving TFMs and updating all the logging frameworks, the only break was log4net's ILoggerRepository dependency.

Yes the NetStandard is quite neat, and now the LibLog makes it very easy to implement logging in any framework. So LibLog is not only logger-independent, it is also framework-independent.

**Comment by damianh** [comment]
> So LibLog is not only logger-independent, it is also framework-independent.

Might use that! Thanks for the help and code review. Much appreciated. :)

**Comment by damianh** [comment]
Mental note: consider supporting the MS.Ext.Logging thing.

**Comment by snakefoot** [comment]
@damianh Maybe make a LibLog release without having updated to Cake/Nuget-package etc. Since implementing a Microsoft-Extension-Logging-Provider is not very close.

So just a standard maintenance release without any NetCore-support besides that it compiles correctly when installed manually.

**Comment by SimonCropp** [comment]
+1 for @snakefoot comment

**Comment by damianh** [comment]
Updated comment. 

Am just going to go to `netstandard2.0` support for v5.0 all the up-to-date platforms support netstandard and net45x is no longer suported. Can consider back port items to a maintenance release.

**Pull request merged**